inspector: use final API

eugeneo opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
deps: new version of the v8_inspector dependency
inspector: updated the code to use the new API


##### Description of change
<!-- Provide a description of the change below this comment. -->
This implementation switches to V8 inspector from the V8 repository. The
new inspector integration is now using final APIs and exposes a stable
wire protocol, removing the need for pointing the users to specific
devtools version.

CC: @ofrobots

eugeneo commented:
I'll be looking into CI failures - CI UI seems to be down, so this might take some time.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -219,8 +214,32 @@ bool RespondToGet(InspectorSocket* socket, const std::string& script_name_,
+        view.length());
+
+  // Reserve 1 character for terminator.
+  std::string result(view.length() * sizeof(UChar) + 1, '\0');
```
`sizeof(UChar)` doesn't look self-evidently correct.  I assume it's shorthand for `sizeof(*view.characters16())` (which is `sizeof(uint16_t)`) but code points >= 0x800 encode to more than two bytes.

eugeneo replied:
Thanks. I changed the sizeof & allowed the code to resize the buffer if need be.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -325,22 +345,24 @@ class DispatchOnInspectorBackendTask : public v8::Task {
  private:
-  void sendProtocolResponse(int callId, const String16& message) override {
+  void sendProtocolResponse(int callId,
+      const v8_inspector::StringView& message) override {
```
Style nit: can you line break at the `(` or line up the second argument?

eugeneo replied:
There's no longer need for a line break here.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -362,8 +384,10 @@ class V8NodeInspector : public v8_inspector::V8InspectorClient {
+    const char CONTEXT_NAME[] = "NodeJS Main Context";
+    inspector_->contextCreated(v8_inspector::V8ContextInfo(env->context(), 1,
+        StringView(reinterpret_cast<const uint8_t*>(CONTEXT_NAME),
+            sizeof(CONTEXT_NAME) - 1)));
```
Style nit: can you line up arguments more here?

eugeneo replied:
I rearranged the code.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -575,15 +599,16 @@ void AgentImpl::InstallInspectorOnProcess() {
+  size_t len = string_value->Length();
+  std::basic_string<uint16_t> buffer(len, '\0');
+  string_value->Write(&buffer[0], 0, len);
+  return StringBuffer::create(StringView(buffer.c_str(), len));
```
For my own curiosity, is `.c_str()` rather than `.data()` necessary here if you also pass `len`?

eugeneo replied:
Coding mistake.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -659,23 +687,24 @@ void AgentImpl::OnRemoteDataIO(InspectorSocket* socket,
+    if (wait_) {
+      std::string message(buf->base, read);
+      if (message.find("\"Runtime.runIfWaitingForDebugger\"")
+          != std::string::npos) {
```
Can you put the operator on the line before?

eugeneo replied:
Done

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -758,8 +788,9 @@ void AgentImpl::SwapBehindLock(MessageQueue* vector1, MessageQueue* vector2) {
-  if (AppendMessage(&incoming_message_queue_, frontend_session_id_, message)) {
+void AgentImpl::PostIncomingMessage(const char* message, size_t len) {
+  if (AppendMessage(&incoming_message_queue_, frontend_session_id_,
+      Utf8ToStringView(message, len))) {
```
Style nit: can you line up the arguments?

eugeneo replied:
Done

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -830,8 +866,9 @@ void AgentImpl::DispatchMessages() {
-  AppendMessage(&outgoing_message_queue_, session_id, message);
+void AgentImpl::Write(int session_id, const StringView& inspector_message) {
+  AppendMessage(&outgoing_message_queue_, session_id,
+      StringBuffer::create(inspector_message));
```
Ditto.

eugeneo replied:
Done

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -804,14 +835,19 @@ void AgentImpl::DispatchMessages() {
+      std::string tag;
+      if (message.length() <= sizeof(TAG_CONNECT) ||
+          message.length() <= sizeof(TAG_DISCONNECT))
+        tag = StringViewToUtf8(message);
```
It's not exactly wrong but I'd wrap the body in braces for legibility.

eugeneo replied:
Done

eugeneo commented:
@bnoordhuis I addressed your comments, please take another look.

CI: https://ci.nodejs.org/job/node-test-pull-request/4506/

Linux ARM failure: exception in Hudson on one bot.
Free BSD: some tests fail on one bot and run on another. The tests do not start inspector.
SmartOS: CI success, not reported to this PR.

eugeneo commented:
@bnoordhuis Please take another look.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -7,19 +7,15 @@
-
 #include "libplatform/libplatform.h"
 
+#include <unicode/unistr.h>
```
If you're taking a dependency on ICU, then configure should have a `--with-intl=none` -> `--without-inspector` implication.

eugeneo replied:
I modified `configure` - now disabling either ICU or OpenSSL will also disable inspector.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -205,8 +202,33 @@ std::string GenerateID() {
+std::string StringViewToUtf8(const StringView& view) {
+  if (view.is8Bit())
+    return std::string(reinterpret_cast<const char*>(view.characters8()),
+        view.length());
```
Can you line up the arguments and maybe put braces around the block?

eugeneo replied:
Done

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -205,8 +202,33 @@ std::string GenerateID() {
+
+  const uint16_t* source = view.characters16();
+  std::string result(view.length() * sizeof(*source), '\0');
+  UnicodeString utf16(reinterpret_cast<const UChar*>(source), view.length());
```
Is this reinterpret_cast necessary?  If for some reason UChar != uint16_t, then this line should arguably fail to compile instead of silently doing the wrong thing.

eugeneo replied:
Windows uses some different type for UChar (not clear from compiler messages - but probably wchar).

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -205,8 +202,33 @@ std::string GenerateID() {
+    CheckedArrayByteSink sink(&result[0], result.length());
+    utf16.toUTF8(sink);
+    result.resize(sink.NumberOfBytesAppended());
+    done = !sink.Overflowed();
```
Does that mean this function may return partial strings?  How bad is that?

bnoordhuis replied:
Correct me if I'm wrong but CheckedArrayByteSink is instantiated every time with the same arguments in this loop so logically `sink.Overflowed()` is either always true or always false.  That means this loop either isn't (i.e., it's not a loop because it exits after the first iteration) or it's an infinite loop.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -205,8 +202,33 @@ std::string GenerateID() {
+std::unique_ptr<StringBuffer> Utf8ToStringView(const char* source,
+                                               size_t length) {
+  UnicodeString utf16 = UnicodeString::fromUTF8(StringPiece(source, length));
+  StringView view(reinterpret_cast<const uint16_t*>(utf16.getBuffer()),
```
Same question as on line 212: is the reinterpret_cast necessary?

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -350,8 +373,10 @@ class V8NodeInspector : public v8_inspector::V8InspectorClient {
                     inspector_(V8Inspector::create(env->isolate(), this)) {
-    inspector_->contextCreated(
-        v8_inspector::V8ContextInfo(env->context(), 1, "NodeJS Main Context"));
+    const uint8_t CONTEXT_NAME[] = "NodeJS Main Context";
```
While you're here: s/NodeJS/Node.js/

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -583,23 +609,25 @@ void AgentImpl::FatalException(v8::Local<v8::Value> error,
-      String16::fromInteger(script_id) == stack_trace->topScriptId()) {
+  v8::Local<v8::StackTrace> stack_trace = message->GetStackTrace();
+
+  if (!stack_trace.IsEmpty() && stack_trace->GetFrameCount() &&
```
Can you put each clause on its own line?

eugeneo replied:
Done

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -815,14 +846,20 @@ void AgentImpl::DispatchMessages() {
+      StringView message = pair.second->string();
+      std::string tag;
+      if (message.length() <= sizeof(TAG_CONNECT) ||
+          message.length() <= sizeof(TAG_DISCONNECT)) {
```
Just `==`?

eugeneo replied:
Done

eugeneo commented:
@bnoordhuis Thank you for the review. I've updated the code, please take another look.

eugeneo commented:
@bnoordhuis Please, take another look.

bnoordhuis commented on configure:
```diff
@@ -1220,6 +1219,11 @@ def configure_intl(o):
+def configure_inspector(o):
+  o['variables']['v8_inspector'] = b(not options.without_inspector
+                                         and o['variables']['v8_enable_i18n_support']
+                                         and o['variables']['node_use_openssl'] == 'true')
```
Can you use `options` here?  That way it won't depend on the order in which the configure_intl an/configure_openssl/configure_inspector functions are executed.

Also, please keep lines <= 80 columns.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -556,23 +584,26 @@ void AgentImpl::FatalException(v8::Local<v8::Value> error,
+  v8::Local<v8::StackTrace> stack_trace = message->GetStackTrace();
+
+  if (!stack_trace.IsEmpty() &&
+      stack_trace->GetFrameCount() &&
```
Can you write this as `stack_trace->GetFrameCount() > 0`?

eugeneo commented:
@bnoordhuis thank you for the review, I uploaded a new version - please take a look.

GitHub Ui is not allowing me to answer this comment (does not show the reply entry field for some reason):

> Correct me if I'm wrong but CheckedArrayByteSink is instantiated every time with the same arguments in this loop so logically sink.Overflowed() is either always true or always false. That means this loop either isn't (i.e., it's not a loop because it exits after the first iteration) or it's an infinite loop.

The code was relying on string length, that was updated in the loop. I changed the code to make it more obvious.

bnoordhuis commented on configure:
```diff
@@ -1220,6 +1219,12 @@ def configure_intl(o):
+def configure_inspector(o):
+  disable_inspector = (options.without_inspector
+                       or options.with_intl in (None, 'none')
+                       or options.without_ssl)
```
I realize this is nitpicking but since we put operators at the end of the line everywhere else can you do it here too?

eugeneo replied:
Done

bnoordhuis commented on configure:
```diff
@@ -1250,6 +1255,8 @@ configure_v8(output)
 configure_openssl(output)
 configure_intl(output)
 configure_static(output)
+# should come after configure_openssl and configure_intl
```
This comment is no longer necessary.

eugeneo replied:
Done.

bnoordhuis commented on src/inspector_agent.cc:
```diff
@@ -177,8 +175,36 @@ const std::string to_string(uint64_t number) {
+  const uint16_t* source = view.characters16();
+  size_t result_length = view.length() * sizeof(*source);
+  std::string result(result_length, '\0');
+  UnicodeString utf16(reinterpret_cast<const UChar*>(source), view.length());
```
If the reinterpret_cast is non-optional here, can you at least add a static_assert that checks the pointed-to types are of the same size?  See https://github.com/nodejs/node/pull/9280 for an example of what I mean.

eugeneo replied:
Done.

eugeneo commented:
Thank you for the review, I addressed the comments and submitted to CI: https://ci.nodejs.org/job/node-test-pull-request/4674/

eugeneo commented:
Landed as 8f00455c5194154ce909ddac6488ae2e42976a4c

TheAlphaNerd commented:
hey @eugeneo this seemed to break configure on v6.x

Would you be willing to manually backport

eugeneo commented:
Let me take a look.

On Fri, Nov 18, 2016 at 1:50 PM Myles Borins <notifications@github.com>
wrote:

> hey @eugeneo <https://github.com/eugeneo> this seemed to break configure
> on v6.x
>
> Would you be willing to manually backport
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/9028#issuecomment-261651112>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARkrQ5CyRRfH7_y8ajVFRyc6bx-Fwtaks5q_h2kgaJpZM4KT2Fh>
> .
>

eugeneo commented:
@TheAlphaNerd backporting this inspector to V8 5.1 is a significant effort. Can we opt out of it? I am not familiar with the process.

TheAlphaNerd commented:
@eugeneo we absolutely can opt out. Considering that v6.x has over 2 years of support left I'd like to get something as up to date as possible though. At what point will we no longer be able to update the inspector that we ship in v6?

eugeneo commented:
I guess this is the point inspector stops getting updated... The new
inspector that is in Node7 is the one imported from the V8 repository (as
opposed to the Blink repository) so it relies on new V8 build setup and
probably some APIs. In the past we had to do some extra work to support
older V8 that the Node uses - but that was lost with the code moving to the
V8 repository.

On Fri, Nov 18, 2016 at 4:08 PM Myles Borins <notifications@github.com>
wrote:

> @eugeneo <https://github.com/eugeneo> we absolutely can opt out.
> Considering that v6.x has over 2 years of support left I'd like to get
> something as up to date as possible though. At what point will we no longer
> be able to update the inspector that we ship in v6?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/9028#issuecomment-261674343>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARkrbf_GPJxY3-YJmNzvjVy_ixxidKpks5q_j3IgaJpZM4KT2Fh>
> .
>
